examples/nfct_test.rs

```diff
@@ -5,166 +5,194 @@ extern crate lang_items;
-                                    bench_transmit(&mut console, &timer, "TX: CC", &mut reply);
-                                }
-                                _ => {
-                                    let mut reply = [rx_buf[0], 0x90, 0x00];
-                                    bench_transmit(&mut console, &timer, "TX: 0x9000", &mut reply);
-                                }
-                            }
-                            _ => {
-                                let mut reply = [rx_buf[0], 0x90, 0x00];
-                                bench_transmit(&mut console, &timer, "TX: 0x9000", &mut reply);
-                            }
-                        }
-                        0x52 | 0x50 /* WUPA | Halt */ => {
-                            if NfcTag::disable_emulation() {
-                                writeln!(console, " -- TAG DISABLED").unwrap();
-                            }
-                            state = State::Disabled;
+    pub fn nfc(mut console: &mut Console) {
+        // Setup the timer with a dummy callback (we only care about reading the current time, but the
+        // API forces us to set an alarm callback too).
+        let mut with_callback = timer::with_callback(|_, _| {});
+        let timer = with_callback.init().flex_unwrap();
+
+        writeln!(
+            console,
+            "Clock frequency: {} Hz",
+            timer.clock_frequency().hz()
+        )
+        .unwrap();
+
+        let mut is_enabled = false;
+        let mut state_change_counter = 0;
+        loop {
+            match is_enabled {
+                true => {
+                    let mut rx_buf = [0; 256];
+                    loop {
+                        receive_packet(&mut console, &mut rx_buf);
+                        // If the reader restarts the communication and we can't
+                        // reply to the received packet, then disable the tag.
+                        if !transmit_reply(&mut console, &timer, &rx_buf) {
+                            is_enabled = false;
                             break;
                         }
-                        _ => (),
                     }
                 }
-            }
-            State::Disabled => {
-                NfcTag::enable_emulation();
-                // Configure Type 4 tag
-                if NfcTag::configure(4) {
-                    state = State::Enabled;
+                false => {
+                    NfcTag::enable_emulation();
+                    // Configure Type 4 tag
+                    if NfcTag::configure(4) {
+                        is_enabled = true;
+                    }
                 }
             }
-        }
-        state_change_cntr += 1;
-        if state_change_cntr > 100 && state == State::Disabled {
-            break;
+            state_change_counter += 1;
+            if state_change_counter > 100 && is_enabled == false {
+                break;
+            }
```

Comment:
Checking into your control flow again, please let me know if this is correct. You are doing the following (shortened):

counter = 0;
> loop {
>   while !configure(4) {
>     enable_emulation();
>     counter += 1;
>   }
>   while !transmit_reply() && counter != 100 {
>     receive_packet();
>     counter += 1;
>   }
> }

If that's roughly the case, you could consider this alternative way to implement this flow. It moves the conditions for the loops into it's head. It reads more naturally than `while(True) break` constructions. Also, it is more obvious that you are alternating between 2 loops, starting with the disabled branch, without the outer `bool` to control that.

Reply:
The first `while` is somewhat correct to what should happen, I did something similar in the upcoming commit. The second loop won't work as the `transmit` check can't be evaluated before receiving first.